File: JavaMVVM/app/src/main/java/com/example/javamvvm/ui/viewmodel/UserViewModel.java
package com.example.javamvvm.ui.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.javamvvm.data.model.User;
import com.example.javamvvm.data.repository.UserRepository;

import java.util.ArrayList;
import java.util.List;

public class UserViewModel extends AndroidViewModel {
    
    private UserRepository repository;
    private LiveData<List<User>> allUsers;
    private MutableLiveData<String> loadingStatus = new MutableLiveData<>();
    private MutableLiveData<String> errorMessage = new MutableLiveData<>();
    
    public UserViewModel(@NonNull Application application) {
        super(application);
        repository = new UserRepository(application);
        allUsers = repository.getAllUsers();
        
        // 初始化示例数据
        loadSampleData();
    }
    
    public LiveData<List<User>> getAllUsers() {
        return allUsers;
    }
    
    public LiveData<String> getLoadingStatus() {
        return loadingStatus;
    }
    
    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }
    
    public void insertUser(User user) {
        repository.insert(user);
    }
    
    public void updateUser(User user) {
        repository.update(user);
    }
    
    public void deleteUser(User user) {
        repository.delete(user);
    }
    
    private void loadSampleData() {
        loadingStatus.setValue("Loading...");
        
        List<User> sampleUsers = new ArrayList<>();
        sampleUsers.add(new User(1, "张三", "zhangsan@example.com", ""));
        sampleUsers.add(new User(2, "李四", "lisi@example.com", ""));
        sampleUsers.add(new User(3, "王五", "wangwu@example.com", ""));
        
        repository.insertAll(sampleUsers);
        loadingStatus.setValue("Loaded");
    }
    
    public void refreshUsers() {
        loadingStatus.setValue("Refreshing...");
        // 这里可以添加网络请求逻辑
        loadingStatus.setValue("Refreshed");
    }
} 

File: KotlinCompose/app/src/main/java/com/example/kotlincompose/presentation/viewmodel/UserViewModel.kt
package com.example.kotlincompose.presentation.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.kotlincompose.domain.model.User
import com.example.kotlincompose.domain.repository.UserRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

data class UserUiState(
    val users: List<User> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null,
    val searchQuery: String = ""
)

@HiltViewModel
class UserViewModel @Inject constructor(
    private val repository: UserRepository
) : ViewModel() {
    
    private val _uiState = MutableStateFlow(UserUiState())
    val uiState: StateFlow<UserUiState> = _uiState.asStateFlow()
    
    private val _searchQuery = MutableStateFlow("")
    
    init {
        loadSampleData()
        observeUsers()
    }
    
    private fun observeUsers() {
        viewModelScope.launch {
            _searchQuery
                .debounce(300)
                .distinctUntilChanged()
                .flatMapLatest { query ->
                    if (query.isBlank()) {
                        repository.getAllUsers()
                    } else {
                        repository.searchUsers(query)
                    }
                }
                .catch { throwable ->
                    _uiState.value = _uiState.value.copy(
                        error = throwable.message,
                        isLoading = false
                    )
                }
                .collect { users ->
                    _uiState.value = _uiState.value.copy(
                        users = users,
                        isLoading = false,
                        error = null
                    )
                }
        }
    }
    
    fun searchUsers(query: String) {
        _searchQuery.value = query
        _uiState.value = _uiState.value.copy(searchQuery = query)
    }
    
    fun addUser(name: String, email: String) {
        viewModelScope.launch {
            try {
                val newUser = User(
                    id = System.currentTimeMillis().toInt(),
                    name = name,
                    email = email,
                    isOnline = (0..1).random() == 1
                )
                repository.insertUser(newUser)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message)
            }
        }
    }
    
    fun updateUser(user: User) {
        viewModelScope.launch {
            try {
                repository.updateUser(user)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message)
            }
        }
    }
    
    fun deleteUser(user: User) {
        viewModelScope.launch {
            try {
                repository.deleteUser(user)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message)
            }
        }
    }
    
    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }
    
    private fun loadSampleData() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            
            val sampleUsers = listOf(
                User(1, "张三", "zhangsan@example.com", "", true),
                User(2, "李四", "lisi@example.com", "", false),
                User(3, "王五", "wangwu@example.com", "", true),
                User(4, "赵六", "zhaoliu@example.com", "", false),
                User(5, "孙七", "sunqi@example.com", "", true)
            )
            
            try {
                repository.insertUsers(sampleUsers)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    error = e.message,
                    isLoading = false
                )
            }
        }
    }
} 